source/v5/mqtt5_to_mqtt3_adapter.c

```diff
@@ -2184,89 +2184,109 @@ static void s_aws_mqtt5_to_mqtt3_adapter_subscribe_completion_fn(
     const struct aws_mqtt5_packet_suback_view *suback,
     int error_code,
     void *complete_ctx) {
-    (void)suback;
 
     struct aws_mqtt5_to_mqtt3_adapter_operation_subscribe *subscribe_op = complete_ctx;
     struct aws_mqtt_client_connection_5_impl *adapter = subscribe_op->base.adapter;
 
-    if (subscribe_op->on_suback != NULL) {
-        AWS_LOGF_DEBUG(
-            AWS_LS_MQTT5_TO_MQTT3_ADAPTER,
-            "id=%p: mqtt3-to-5-adapter, completing single-topic subscribe",
-            (void *)adapter);
+    if (suback == NULL) {
+        if (subscribe_op->on_suback != NULL) {
+            AWS_LOGF_DEBUG(
+                AWS_LS_MQTT5_TO_MQTT3_ADAPTER,
+                "id=%p: mqtt3-to-5-adapter, completing single-topic subscribe",
+                (void *)adapter);
+            (*subscribe_op->on_suback)(
+                &adapter->base,
+                subscribe_op->base.id,
+                NULL,
+                AWS_MQTT_QOS_FAILURE,
+                error_code,
+                subscribe_op->on_suback_user_data);
+        }
+        if (subscribe_op->on_multi_suback != NULL) {
+            AWS_LOGF_DEBUG(
+                AWS_LS_MQTT5_TO_MQTT3_ADAPTER,
+                "id=%p: mqtt3-to-5-adapter, completing multi-topic subscribe",
+                (void *)adapter);
+            (*subscribe_op->on_multi_suback)(
+                &adapter->base, subscribe_op->base.id, NULL, error_code, subscribe_op->on_multi_suback_user_data);
```

Comment:
`aws-crt-cpp` expects the list to be non-empty. It won't crash, but it'll set the qos to `AWS_MQTT_QOS_AT_MOST_ONCE` and then pass empty list of topics to the user callback. Which may be confusing.
Here's the corresponding code: https://github.com/awslabs/aws-crt-cpp/blob/main/source/mqtt/MqttConnectionCore.cpp#L414

Probably the empty list should be handled in `aws-crt-cpp`. But I feel like it should be done here, so all the layers above don't have to worry about these things.

Reply:
I thought about it. I think it might be better to handle it on CRT side, as it might be better to check NULL value for a pointer regardless. 
I will update the changes in CRTs.